File: Popcorn/Domain/Repositories/MovieRepository.swift
//
//  MovieRepository.swift
//  Popcorn
//
//  Created by Eduardo Raupp Peretto on 08/05/25.
//

import Combine

protocol MovieRepository {
    func fetchMovies(filtering: FilterOptions?, page: Int) -> AnyPublisher<PaginatedMovies, Error>
    func fetchMovieDetails(id: Int) -> AnyPublisher<MovieDetails, Error>

    func fetchLikedMovies() -> AnyPublisher<[Movie], Error>
    func storeLikedMovie(_ movie: Movie) -> AnyPublisher<Void, Error>
    func removeLikedMovie(_ movie: Movie) -> AnyPublisher<Void, Error>

    func fetchWatchLaterMovies() -> AnyPublisher<[Movie], Error>
    func storeWatchLaterMovie(_ movie: Movie) -> AnyPublisher<Void, Error>
    func removeWatchLaterMovie(_ movie: Movie) -> AnyPublisher<Void, Error>
}

File: KotlinShared/shared/src/commonMain/kotlin/com/eduardoperetto/popcornshared/domain/MovieRepository.kt
package com.eduardoperetto.popcornshared.domain

// Repository Interface
interface MovieRepository {
    suspend fun fetchMovies(filtering: FilterOptions?, page: Int): Response<PaginatedMovies>
    suspend fun fetchMovieDetails(id: Int): Response<MovieDetails>

    suspend fun fetchLikedMovies(): Response<List<Movie>>
    suspend fun storeLikedMovie(movie: Movie): Response<Unit>
    suspend fun removeLikedMovie(movie: Movie): Response<Unit>

    suspend fun fetchWatchLaterMovies(): Response<List<Movie>>
    suspend fun storeWatchLaterMovie(movie: Movie): Response<Unit>
    suspend fun removeWatchLaterMovie(movie: Movie): Response<Unit>
}